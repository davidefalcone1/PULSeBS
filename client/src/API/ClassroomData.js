class ClassroomData{
    constructor(id, classroomName, maxSeats) {
        this.classId = id;
        this.classroomName = classroomName;
        this.maxSeats = maxSeats;
    }

    static fromJson(json){
        const temp =  Object.assign(new ClassroomData(), json);
        return temp;
    }
}

export default ClassroomData;