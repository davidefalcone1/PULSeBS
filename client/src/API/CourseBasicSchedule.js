class CourseBasicSchedule{
    constructor(id, courseId, day, startTime, endTime, classroom) {
        this.id = id;
        this.courseId = courseId;
        this.day = day;
        this.startTime = startTime;
        this.endTime = endTime;
        this.classroom = classroom;
    }

    static fromJson(json){
        const temp =  Object.assign(new CourseBasicSchedule(), json);
        return temp;
    }
}

export default CourseBasicSchedule;