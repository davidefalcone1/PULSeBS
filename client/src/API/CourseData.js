//usata per la retrieve della lista dei corsi di un docente
class CourseData{
    constructor(courseId, courseName, teacherId, 
        normalBookingsAvgWeek, cancelledBookingsAvgWeek, waitingBookingsAvgWeek, 
        normalBookingsAvgMonth, cancelledBookingsAvgMonth, waitingBookingsAvgMonth,
        attendanceCountAvgMonth, attendanceCountAvgWeek ) {
        if (courseId) 
            this.courseId = courseId;
        this.courseName = courseName;
        this.teacherId = teacherId;
        this.normalBookingsAvgWeek = normalBookingsAvgWeek;
        this.cancelledBookingsAvgWeek = cancelledBookingsAvgWeek;
        this.waitingBookingsAvgWeek = waitingBookingsAvgWeek;
        this.normalBookingsAvgMonth = normalBookingsAvgMonth;
        this.cancelledBookingsAvgMonth = cancelledBookingsAvgMonth;
        this.waitingBookingsAvgMonth = waitingBookingsAvgMonth;
        this.attendanceCountAvgWeek = attendanceCountAvgWeek;
        this.attendanceCountAvgMonth = attendanceCountAvgMonth;
    }

    static fromJson(json){
        const temp =  Object.assign(new CourseData(), json);
        return temp;
    }
}

export default CourseData;