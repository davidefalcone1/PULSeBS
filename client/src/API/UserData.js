class UserData{
    constructor(id, personId, fullName, email) {
        if (id) 
            this.id = id;
        this.personId = personId;
        this.fullName = fullName;
        this.email = email;
    }

    static fromJson(json){
        const temp =  Object.assign(new UserData(), json);
        return temp;
    }
}

export default UserData;