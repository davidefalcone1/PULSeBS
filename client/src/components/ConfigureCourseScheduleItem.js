import React from 'react';
import ListGroup from 'react-bootstrap/ListGroup';
import Button from 'react-bootstrap/Button';

const configureCourseScheduleItem = (props) => {

  return (
    <ListGroup.Item id = {props.schedule.id + "-" + props.schedule.scheduleId}>
        <div className="d-flex w-100 pt-3 justify-content-between no-gutters">
            <DayField id = {props.schedule.id} day = {props.schedule.day}/>
            <StartTimeField id = {props.schedule.id} startTime = {props.schedule.startTime}/>
            <EndTimeField id = {props.schedule.id} endTime = {props.schedule.endTime}/>
            <ButtonField id = {props.schedule.id} edit = {props.editCourseSchedule}
                delete = {props.deleteCourseSchedule}/>
        </div>
    </ListGroup.Item>
  );
}

function DayField(props){
    return(
        <div className="col-sm-3">
            <h4 id={"dayOfSchedule" + props.id}>
                {props.day}
            </h4>
        </div>
    );
}

function StartTimeField(props){
    return(
        <div className="col-sm-3">
            <p id={"sTimeOfSchedule" + props.id}>
                {props.startTime}
            </p>
        </div>
    );
}

function EndTimeField(props){
    return(
        <div className="col-sm-3">
            <p id={"eTimeOfSchedule" + props.id}>
                {props.endTime}
            </p>
        </div>
    );
}
function ButtonField(props){
    return(
        <div className="col-sm-4">
            <Button variant="info" onClick={(event) => {
                event.preventDefault();
                props.edit();
            }} id={"editButtonFieldOflesson" + props.id}>
                Edit
            </Button>
            <Button variant="danger" onClick={(event) => {
                event.preventDefault();
                props.delete(props.id);
            }} id={"deleteButtonFieldOflesson" + props.id}>
                Delete
            </Button>
        </div>
    );
}

export default configureCourseScheduleItem;