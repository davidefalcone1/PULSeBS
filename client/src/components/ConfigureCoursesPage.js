import React from 'react';
import ConfigureCourseItem from './ConfigureCoursesItem';
import ListGroup from 'react-bootstrap/ListGroup';
import Button from 'react-bootstrap/esm/Button';
import Row from 'react-bootstrap/Row';
import Modal from 'react-bootstrap/Modal';
import Alert from 'react-bootstrap/Alert';
import Form from 'react-bootstrap/Form';
import { AuthContext } from '../_services/AuthContext';
import { Redirect } from 'react-router-dom';

const configureCoursePage = (props) => {
  return(
    <ConfigureCourses coursesList={props.coursesList} createNewCourse={props.createNewCourse}
      teachersList={props.teachersList} uploadFileCourses={props.uploadFileCourses}/>
  );
}

class ConfigureCourses extends React.Component {
  constructor(props) {
    super(props);
    this.props = props;
    this.state = {
      file: undefined, isUploading: false, errorFile: false,
      isCreating: false, courseName: '', teacherId: 'Select teacher', 
      errorName: false, errorTeacher: false
    }
  }

  activateModal = () => {
    this.setState({isCreating: true});
  }
  activateUploadFileModal = () => {
    this.setState({isUploading: true})
  }

  updateField = (name, value) => {
    this.setState({[name]: value}, () => {
      if(this.state.courseName !== 'Select teacher'){
        this.setState({errorName: false});
      }
      if(this.state.teacherId !== 'Select teacher'){
          this.setState({errorTeacher: false});
      }
      if(this.state.file !== undefined && this.state.file !== ''){
        this.setState({errorFile: false});
      }
    });
  }

  handleSubmit = () => {
    if (!this.form.checkValidity()) {
        this.form.reportValidity();
    }
    else if(this.state.courseName === 'Select teacher'){
        this.setState({errorName: true});
    }
    else if(this.state.teacherId === 'Select teacher'){
        this.setState({errorTeacher: true});
    }
    else {
        this.props.createNewCourse(this.state.courseName, this.state.teacherId)
    }
  }
  handleSubmitFile = () => {
    if (!this.formFile.checkValidity()) {
      this.formFile.reportValidity();
    }
    else if(this.state.file === undefined || this.state.file === ''){ 
      this.setState({errorFile: true});
    }
    else {
        this.props.uploadFileCourses(this.state.file)
        this.setState({isUploading: false})
    }
  }

  render(){
    return(
      <AuthContext.Consumer>
        {(context) => (
          <>        
            {context.user && this.props.coursesList && this.props.teachersList &&
              <>
                <br/>
                <Row className="justify-content-around">
                    <Button variant="primary" onClick={(event) => {
                          event.preventDefault();
                          this.activateModal();
                      }} id={"activateModalOfCourses"}>
                          Add New
                    </Button>
                    <Button variant="primary" onClick={(event) => {
                          event.preventDefault();
                          this.activateUploadFileModal();
                      }} id={"uploadFileOfCourses"}>
                          Upload File
                    </Button>
                </Row>
                <ListGroup as="ul" variant="flush">
                  <ListHeader />
                  {this.props.coursesList.map((course) => 
                      <ConfigureCourseItem key = {course.courseId} course = {course} 
                          teachers = {this.props.teachersList}/>)
                  }
                </ListGroup>

                <Modal show={this.state.isCreating} animation={false} scrollable={true}>
                  <Modal.Header>
                    <Modal.Title>Create new course</Modal.Title>
                  </Modal.Header>
                  <Modal.Body>
                    <Form method="POST" action="" id="newCourseForm" onSubmit={(ev) => {
                      ev.preventDefault();
                      this.handleSubmit();
                    }} ref={(form) => this.form = form}>
                          
                      <Form.Group>
                        <Form.Label className="control-label">Course Name</Form.Label>
                        <Form.Control type="text" name="courseName" size = "lg"
                          value = {this.state.classRoomName} required autoFocus
                          onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}/>
                      </Form.Group>
                      <Form.Group>
                        <Form.Label className="control-label">Teacher</Form.Label>
                        <Form.Control as="select" custom name="teacherId" value = {this.state.courseName} 
                            onChange={(ev) => this.updateField(ev.target.name, ev.target.value)}>
                          <option>Select teacher</option>
                          {this.props.teachersList.map((teacher) =>
                            <option value={teacher.personId}>{teacher.fullName}</option>
                          )}
                        </Form.Control>
                      </Form.Group>
                      <Form.Group>
                        <div>
                          <button type="submit" className="btn btn-primary">Create</button>
                        </div>
                      </Form.Group>
                    </Form>
                  </Modal.Body>
                  <Modal.Footer>
                    <Button variant="secondary" type="button" 
                      onClick={(event) => {
                        event.preventDefault();
                        this.setState({isCreating: false});
                      }}>Close</Button>
                      {this.state.errorName &&
                        <>
                          <br/>
                          <Alert key="nameError" variant="danger">
                            Invalid name.
                          </Alert>
                        </>}
                      {this.state.errorTeacher &&
                        <>
                          <br/>
                          <Alert key="teacherError" variant="danger">
                            Invalid teacher.
                          </Alert>
                        </>}
                  </Modal.Footer>
                </Modal>
                
                <Modal show={this.state.isUploading} animation={false} scrollable={true}>
                  <Modal.Header>
                    <Modal.Title>Upload file</Modal.Title>
                  </Modal.Header>
                  <Modal.Body>
                    <Form method="POST" action="" id="newCourseFormFile" onSubmit={(ev) => {
                      ev.preventDefault();
                      this.handleSubmitFile();
                    }} ref={(form) => this.formFile = form}>                          
                      <Form.Group>
                        <Form.Label className="control-label">Insert file</Form.Label>
                        <Form.Control type="file" name="file" size = "lg"
                          required autoFocus accept=".csv"
                          onChange={(ev) => {                            
                            var f2 =function readFileContent(file) {
                              const reader = new FileReader()
                                return new Promise((resolve, reject) => {
                                  reader.onload=event=>resolve(event.target.result)
                                  reader.onerror = error => reject(error)
                                  reader.readAsText(file)
                                })
                            }
                            
                            f2(ev.target.files[0]).then(content => {
                              this.updateField("file", content)
                            }).catch(error => console.log(error))
                          }}/>
                      </Form.Group>
                      <Form.Group>
                        <div>
                          <button type="submit" className="btn btn-primary">
                            UPLOAD
                          </button>
                        </div>
                      </Form.Group>
                    </Form>
                  </Modal.Body>
                  <Modal.Footer>
                    <Button variant="secondary" type="button" 
                      onClick={(event) => {
                        event.preventDefault();
                        this.setState({isUploading: false});
                      }}>Close</Button>
                    {this.state.errorFile &&
                      <>
                        <br/>
                        <Alert key="fileError" variant="danger">
                          Invalid file.
                        </Alert>
                      </>}
                  </Modal.Footer>
                </Modal>
              </>
            }

            {context.user && (!this.props.coursesList || this.props.teachersList) && 
              <NoItemsImage/>
            }
            {!context.user && <Redirect to="/login"/>}
          </>
        )}
      </AuthContext.Consumer>
    );
  };  
}

function ListHeader() {
  return(
    <ListGroup.Item id = {"coursesList-header"}>
        <div className="d-flex w-100 pt-3 justify-content-between no-gutters">
          <div className="col-sm-6">
            <h4>Course Name</h4>
          </div>
          <div className="col-sm-6">
            <h4>Teacher</h4>
          </div>
        </div>
    </ListGroup.Item>
  );
}
function NoItemsImage(props){
  return(
      <div className="col-sm-12 pt-3">
          <img width="800" height="600" className="img-button" src='./images/no_data_image.png' alt=""/>
      </div>
  );
}

export default configureCoursePage;
