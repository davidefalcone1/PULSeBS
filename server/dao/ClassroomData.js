class ClassroomData{
    constructor(id, classroomName, maxSeats) {
        this.classId = id;
        this.classroomName = classroomName;
        this.maxSeats = maxSeats;
    }

}

module.exports = ClassroomData;