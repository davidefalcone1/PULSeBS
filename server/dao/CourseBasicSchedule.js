class CourseBasicSchedule{
    constructor(id, courseId, day, startTime, endTime, classroom) {
        this.id = id;
        this.courseId = courseId;
        this.day = day;
        this.startTime = startTime;
        this.endTime = endTime;
        this.classroom = classroom;
    }
}

module.exports = CourseBasicSchedule;