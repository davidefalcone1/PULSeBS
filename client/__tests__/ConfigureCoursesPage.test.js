import React from 'react';
import ConfigureCoursesPage from '../src/components/ConfigureCoursesPage';
import CourseData from '../src/API/CourseData';
import UserData from '../src/API/UserData';
import { mount } from 'enzyme';
import CourseBasicSchedule from '../src/API/CourseBasicSchedule';
import ConfigureCourseScheduleItem from '../src/components/ConfigureCourseScheduleItem';
import ClassroomData from '../src/API/ClassroomData';
import Accordion from 'react-bootstrap/Accordion';

describe('ConfigureCoursesPage', ()=>{
    const component = mount(
        <ConfigureCoursesPage coursesList={[new CourseData('XY123', 'Software engineering 2', '654321', 2, 3, 5, 6, 7, 3, 5, 7)]} teachersList={[new UserData(1, '654321', 'Mario Rossi', 'mario.rossi@polito.it')]} createNewCourse={()=>{}} uploadFileCourses={()=>{}} basicSchedules={[new CourseBasicSchedule(1, 'XY123', 'Mon', '02:00', '05:00', 'A1')]} editCourseSchedule={()=>{}} createNewCourseSchedule={()=>{}} deleteCourseSchedule={()=>{}} classroomssList={[new ClassroomData(1, 'A1', 90)]}/>,
        {
            context: {user: new UserData(1, '12345', 'Davide Falcone', 'davide.falcone@polito.it')}
        }
    );
    test('Renders', ()=>{
        expect(component).toContainMatchingElement('#XY123');
    });
    test('Add new Modal renders', ()=>{
        component.find('#activateModalOfCourses').at(0).simulate('click'); // open Modal
        component.find('input').at(0).simulate('change', {target: {value: 'My new value'}}); //Type something
        component.find('select').at(0).simulate('change', {target: {value: '654321'}}); //Type something
    });
    test('Upload file Modal renders', ()=>{
        component.find('#uploadFileOfCourses').at(0).simulate('click'); // open modal
    });
    test('Create new schedule Modal renders', (done)=>{
        component.find('.card-header').find('button').at(0).simulate('click'); //open accordion
        setTimeout(()=>{ //wait for accordion to open
            component.find('#buttonFieldOflessonXY123').at(0).simulate('click'); //open modal
            component.find('input').at(0).simulate('change', {target: {value: 'My new value'}}); //Type something
            component.find('input').at(1).simulate('change', {target: {value: 'My new value'}}); //Type something
            component.find('select').at(0).simulate('change', {target: {value: 'Monday'}}); 
            component.find('select').at(1).simulate('change', {target: {value: '1 - 120'}}); 
            component.find('#newCourseForm').at(0).simulate('submit');
            done();
        }, 500);
    });
    test('Open course accordion', (done)=>{
        component.find('.card-header').find('button').at(0).simulate('click'); //open accordion
        setTimeout(()=>{
            expect(component).toContainExactlyOneMatchingElement(ConfigureCourseScheduleItem);
            done();
        }, 500);
    });
});