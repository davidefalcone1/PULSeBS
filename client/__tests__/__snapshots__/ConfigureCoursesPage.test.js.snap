// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigureCoursesPage Renders correctly 1`] = `
Array [
  <br />,
  <div
    className="justify-content-around row"
  >
    <button
      className="btn btn-primary"
      disabled={false}
      id="activateModalOfCourses"
      onClick={[Function]}
      type="button"
    >
      Add New
    </button>
    <button
      className="btn btn-primary"
      disabled={false}
      id="uploadFileOfCourses"
      onClick={[Function]}
      type="button"
    >
      Upload File
    </button>
  </div>,
  <ul
    className="list-group list-group-flush"
    onKeyDown={[Function]}
  >
    <div
      className="list-group-item"
      data-rb-event-key={null}
      disabled={false}
      id="coursesList-header"
      onClick={[Function]}
    >
      <div
        className="d-flex w-100 pt-3 justify-content-between no-gutters"
      >
        <div
          className="col-sm-6"
        >
          <h4>
            Course Name
          </h4>
        </div>
        <div
          className="col-sm-6"
        >
          <h4>
            Teacher
          </h4>
        </div>
      </div>
    </div>
    <div
      className="list-group-item"
      data-rb-event-key={null}
      disabled={false}
      id="course-XY123"
      onClick={[Function]}
    >
      <div
        className="d-flex w-100 pt-3 justify-content-between no-gutters"
      >
        <div
          className="col-sm-6"
        >
          <h6
            id="nameOfcourse_XY123"
          >
            Software engineering 2
          </h6>
        </div>
        <div
          className="col-sm-6"
        >
          <h6
            id="teacherOfcourse_654321-XY123"
          >
            Mario Rossi
          </h6>
        </div>
      </div>
    </div>
  </ul>,
]
`;
